### PR TITLE
Various fixes based on testing on Core

### PR DESCRIPTION
**Pull request by paracycle**
This PR makes a few fixes to ensure everything works properly on Core. I've also taken the opportunity to increase strictness of DSL compilers to `strict`.

It is best to review commit-by-commit

### CODE REVIEW

File: lib/tapioca/compilers/dsl/url_helpers.rb [hunk 1]
```diff
@@ -36,7 +36,7 @@ def gather_constants
 
           constants = ObjectSpace.each_object(Module).select do |mod|
             mod = T.cast(mod, T.class_of(Module))
-            next unless mod.name
+            next unless Module.instance_method(:name).bind(mod).call
```
Comment (katayounHillier) [inline lib/tapioca/compilers/dsl/url_helpers.rb:39]: What sort of issues we are fixing with this change?

File: lib/tapioca/compilers/dsl/smart_properties.rb [hunk 2]
```diff
@@ -127,7 +127,10 @@ def generate_methods_for_property(klass, property, &block)
           klass.create_method(property.reader.to_s, return_type: type) if block.call(property.reader.to_s)
         end
 
-        BOOLEANS = [[true, false], [false, true]].freeze
+        BOOLEANS = T.let([
+          [true, false],
+          [false, true]
+        ].freeze, T::Array[[T::Boolean, T::Boolean]])
```
Comment (katayounHillier) [inline lib/tapioca/compilers/dsl/smart_properties.rb:133]: is this change just for making it more secure? or was there an actual issue that this change is fixing

File: lib/tapioca/compilers/dsl/url_helpers.rb [hunk 3]
```diff
@@ -65,11 +65,13 @@ def generate_module_for(root, constant)
           end
         end
 
+        sig { params(mod: Parlour::RbiGenerator::Namespace, constant: T.class_of(Module), helper_module: Module).void }
         def create_mixins_for(mod, constant, helper_module)
-          mod.create_include(helper_module.name) if constant.ancestors.include?(helper_module)
-          mod.create_extend(helper_module.name) if constant.singleton_class.ancestors.include?(helper_module)
+          mod.create_include(T.must(helper_module.name)) if constant.ancestors.include?(helper_module)
```
Comment (katayounHillier) [inline lib/tapioca/compilers/dsl/url_helpers.rb:70]: What does `T.must` do here?

File: lib/tapioca/compilers/dsl/url_helpers.rb (hunk 1 above)
Comment (paracycle) [reply lib/tapioca/compilers/dsl/url_helpers.rb:39]: If you have a type that redefines the `name` class method, then the value returned from it ends up being useless. For example, if I define a class `Foo`, I would expect `Foo.name == "Foo"` to be true. However, if I defined `Foo` as:
```ruby
class Foo
  def self.name
    nil
  end
end
```
then `Foo.name` would return `nil`, which is not what we want. We want to get the name of the class.

So the solution is to call the `name` instance method of `Module` and call that method on the object that we are working with. So in the above case `Module.instance_method(:name).bind(Foo).call` would correctly return `"Foo"`.

That is what this change is doing.

File: lib/tapioca/compilers/dsl/smart_properties.rb (hunk 2 above)
Comment (paracycle) [reply lib/tapioca/compilers/dsl/smart_properties.rb:133]: No, this is necessary for `typed: strict`. At that strictness level, all constants must have their types declared using `T.let`.

File: lib/tapioca/compilers/dsl/url_helpers.rb (hunk 3 above)
Comment (paracycle) [reply lib/tapioca/compilers/dsl/url_helpers.rb:70]: In general, the `name` method for a random `Module` could return `nil` if the module in question does not have a name. For example:
```ruby
a = Class.new
a.name # => nil
```
So the return type of the `.name` method is `T.nilable(String)`. But `create_include` is expecting a `String` which should not be `nil`. `T.must` tells Sorbet that we expect the return value of `helper_module.name` to **not** be `nil`. So Sorbet stops complaining. 

By the way, if code reaches this method, we **do** know that the module will definitely have a name, so this is safe to do.

**Pull request merged**